### PR TITLE
Merge #198

### PR DESCRIPTION
**Pull request by michielbdejong**
Rebase of #198.

### CODE REVIEW

**Comment by michielbdejong** [comment]
This is how IL-RFC4 in this branch differs from Stefan's version:
```
[michiel@mbp ~/gh/interledger/rfcs (mj-rebase-docs-st-req-res)]$ git diff origin/docs/st-req-res
diff --git a/0004-ledger-plugin-interface/0004-ledger-plugin-interface.md b/0004-ledger-plugin-interface/0004-ledger-plugin-interface.md
index a6861ce..92fa21c 100644
--- a/0004-ledger-plugin-interface/0004-ledger-plugin-interface.md
+++ b/0004-ledger-plugin-interface/0004-ledger-plugin-interface.md
@@ -32,24 +32,24 @@ This spec depends on the [ILP spec](../0003-interledger-protocol/).
 ###### Events
 | Name | Handler |
 |:--|:--|
-| [**connect**](#event-connect-) | `( ) ⇒` |
-| [**disconnect**](#event-disconnect-) | `( ) ⇒` |
-| [**error**](#event-error-) | `( ) ⇒` |
-| [**incoming_transfer**](#event-*_transfer-) | <code>( transfer:[Transfer](#transfer) ) ⇒</code> |
-| [**incoming_prepare**](#event-*_prepare-) | <code>( transfer:[Transfer](#transfer) ) ⇒</code> |
-| [**incoming_fulfill**](#event-*_fulfill-) | <code>( transfer:[Transfer](#transfer), fulfillment:String ) ⇒</code> |
-| [**incoming_reject**](#event-*_reject-) | <code>( transfer:[Transfer](#transfer), rejectionReason:[RejectionMessage](#class-rejectionmessage) ) ⇒</code> |
-| [**incoming_cancel**](#event-*_cancel-) | <code>( transfer:[Transfer](#transfer), cancellationReason:[RejectionMessage](#class-rejectionmessage) ) ⇒</code> |
-| [**incoming_request**](#event-*_request-) | <code>( message:[Message](#message) ) ⇒</code> |
-| [**incoming_response**](#event-*_response-) | <code>( message:[Message](#message) ) ⇒</code> |
-| [**outgoing_transfer**](#event-*_transfer-) | <code>( transfer:[Transfer](#transfer) ) ⇒</code> |
-| [**outgoing_prepare**](#event-*_prepare-) | <code>( transfer:[Transfer](#transfer) ) ⇒</code> |
-| [**outgoing_fulfill**](#event-*_fulfill-) | <code>( transfer:[Transfer](#transfer), fulfillment:String ) ⇒</code> |
-| [**outgoing_reject**](#event-*_reject-) | <code>( transfer:[Transfer](#transfer), rejectionReason:[RejectionMessage](#class-rejectionmessage) ) ⇒</code> |
-| [**outgoing_cancel**](#event-*_cancel-) | <code>( transfer:[Transfer](#transfer), cancellationReason:[RejectionMessage](#class-rejectionmessage) ) ⇒</code> |
-| [**outgoing_request**](#event-*_request-) | <code>( message:[Message](#message) ) ⇒</code> |
-| [**outgoing_response**](#event-*_response-) | <code>( message:[Message](#message) ) ⇒</code> |
-| [**info_change**](#event-info_change-) | <code>( info:[LedgerInfo](#class-ledgerinfo) ) ⇒</code> |
+| [**connect**](#event-connect) | `( ) ⇒` |
+| [**disconnect**](#event-disconnect) | `( ) ⇒` |
+| [**error**](#event-error) | `( ) ⇒` |
+| [**incoming_transfer**](#event-_transfer) | <code>( transfer:[IncomingTransfer](#incomingtransfer) ) ⇒</code> |
+| [**incoming_prepare**](#event-_prepare) | <code>( transfer:[IncomingTransfer](#incomingtransfer) ) ⇒</code> |
+| [**incoming_fulfill**](#event-_fulfill) | <code>( transfer:[IncomingTransfer](#incomingtransfer), fulfillment:String ) ⇒</code> |
+| [**incoming_reject**](#event-_reject) | <code>( transfer:[IncomingTransfer](#incomingtransfer), rejectionReason:[RejectionMessage](#class-rejectionmessage) ) ⇒</code> |
+| [**incoming_cancel**](#event-_cancel) | <code>( transfer:[IncomingTransfer](#incomingtransfer), cancellationReason:[RejectionMessage](#class-rejectionmessage) ) ⇒</code> |
+| [**incoming_request**](#event-_request) | <code>( message:[Message](#message) ) ⇒</code> |
+| [**incoming_response**](#event-_response) | <code>( message:[Message](#message) ) ⇒</code> |
+| [**outgoing_transfer**](#event-_transfer) | <code>( transfer:[outgoingTransfer](#outgoingtransfer) ) ⇒</code> |
+| [**outgoing_prepare**](#event-_prepare) | <code>( transfer:[outgoingTransfer](#outgoingtransfer) ) ⇒</code> |
+| [**outgoing_fulfill**](#event-_fulfill) | <code>( transfer:[outgoingTransfer](#outgoingtransfer), fulfillment:String ) ⇒</code> |
+| [**outgoing_reject**](#event-_reject) | <code>( transfer:[outgoingTransfer](#outgoingtransfer), rejectionReason:[RejectionMessage](#class-rejectionmessage) ) ⇒</code> |
+| [**outgoing_cancel**](#event-_cancel) | <code>( transfer:[outgoingTransfer](#outgoingtransfer), cancellationReason:[RejectionMessage](#class-rejectionmessage) ) ⇒</code> |
+| [**outgoing_request**](#event-_request) | <code>( message:[Message](#message) ) ⇒</code> |
+| [**outgoing_response**](#event-_response) | <code>( message:[Message](#message) ) ⇒</code> |
+| [**info_change**](#event-info_change) | <code>( info:[LedgerInfo](#class-ledgerinfo) ) ⇒</code> |

 ###### Errors
 | Name | Description |
@@ -343,7 +343,7 @@ to you.

 Emitted when an outgoing/incoming transfer containing a condition is prepared.

-Note that the `*_prepare` event **DOES NOT** indicate that money has been transferred. The final status will only be known when either the [*_fulfill](#event-*_fulfill-) or [*_cancel](#event-*_cancel-) events are emitted.
+Note that the `*_prepare` event **DOES NOT** indicate that money has been transferred. The final status will only be known when either the [*_fulfill](#event-_fulfill) or [*_cancel](#event-_cancel) events are emitted.

 The ledger plugin MUST authenticate the source for all incoming transfers, whether they include money or not.

@@ -638,9 +638,11 @@ Metadata describing the ledger. This data is returned by the [`getInfo`](#getinf
 | Type | Name | Description |
 |:--|:--|:--|
 | `String` | [prefix](#ledgerinfo-prefix) | The plugin's ILP address prefix |
-| `String` | [currencyCode](#ledgerinfo-currencyCode) | [ISO 4217](https://en.wikipedia.org/wiki/ISO_4217) three-letter currency code |
-| `Number` | [currencyScale](#ledgerinfo-currencyScale) | Integer `(..., -2, -1, 0, 1, 2, ...)`, such that one of the ledger's base units equals `10^-<currencyScale> <currencyCode>` |
+| `String` | [currencyCode](#ledgerinfo-currencycode) | [ISO 4217](https://en.wikipedia.org/wiki/ISO_4217) three-letter currency code |
+| `Number` | [currencyScale](#ledgerinfo-currencyscale) | Integer `(..., -2, -1, 0, 1, 2, ...)`, such that one of the ledger's base units equals `10^-<currencyScale> <currencyCode>` |
 | `String[]` | [connectors](#ledgerinfo-connectors) | ILP addresses of recommended connectors |
+| `String` | [minBalance](#ledgerinfo-minbalance-optional) | Integer String, for instance `"0"`, indicating the minimum balance. Optional, defaults to zero. |
+| `String` | [maxBalance](#ledgerinfo-maxbalance-optional) | Integer String, for instance `"1000000000000"`, indicating the maximum balance. Optional, defaults to plus infinity. |

 ### Fields

@@ -672,6 +674,24 @@ currencyCode is `USD` and its currencyScale is `2`.

 The ILP addresses of recommended connectors.

+#### LedgerInfo#minBalance (OPTIONAL)
+<code>**minBalance**:String</code>
+
+A minimum balance limits how much the ledger trusts the account holder.
+This field is optional; when not present, the minimum balance should be assumed to be `0`.
+When a plugin does return a `minBalance` field, it should be an Integer String, measured in the ledger's base unit,
+comparable to the `balance` Integer Strings for which the `getBalance` method returns a Promise.
+Applications using the plugin can expect transfers to fail if they would make the balance go below the minimum.
+
+#### LedgerInfo#maxBalance (OPTIONAL)
+<code>**maxBalance**:String</code>
+
+A maximum balance limits how much the account holder trusts the ledger.
+This field is optional; when not present, the maximum balance should be assumed to be `+Infinity`.
+When a plugin does return a `maxBalance` field, it should be an Integer String, measured in the ledger's base unit,
+When a plugin does return a `maxBalance` field, it should be an Integer String, measured in the ledger's base unit,
+comparable to the `balance` Integer Strings for which the `getBalance` method returns a Promise.
+Applications using the plugin can expect transfers to fail if they would make the balance exceed the maximum.
+
 ## Class: PluginOptions
 <code>class PluginOptions</code>
```

So I think that's a correct rebase.

**Comment by michielbdejong** [comment]
This is the meta-diff, that's to say, the diff between how master has changed and how my rebase changes the PR:
```diff
[michiel@mbp ~/gh/interledger/rfcs (master)]$ diff old-diff.txt new-diff.txt
2c2
< index 67462da..23603a8 100644
---
> index a6861ce..92fa21c 100644
5,15c5
< @@ -22,29 +22,29 @@ This spec depends on the [ILP spec](../0003-interledger-protocol/).
<  | | [**getAccount**](#getaccount) ( ) `⇒ String` |
<  | | [**getBalance**](#getbalance) ( ) <code>⇒ Promise.&lt;String></code> |
<  | | [**getFulfillment**](#getfulfillment) ( transferId ) <code>⇒ Promise.&lt;String></code> |
< -| | [**sendTransfer**](#sendTransfer) ( transfer ) <code>⇒ Promise.&lt;null></code> |
< -| | [**sendMessage**](#sendMessage) ( message ) <code>⇒ Promise.&lt;null></code> |
< +| | [**sendTransfer**](#sendtransfer) ( transfer ) <code>⇒ Promise.&lt;null></code> |
< +| | [**sendMessage**](#sendmessage) ( message ) <code>⇒ Promise.&lt;null></code> |
<  | | [**fulfillCondition**](#fulfillcondition) ( transferId, fulfillment ) <code>⇒ Promise.&lt;null></code> |
<  | | [**rejectIncomingTransfer**](#rejectincomingtransfer) ( transferId, reason ) <code>⇒ Promise.&lt;null></code> |
<
---
> @@ -32,24 +32,24 @@ This spec depends on the [ILP spec](../0003-interledger-protocol/).
22,32c12,25
< -| [**incoming_transfer**](#event-*_transfer-) | <code>( transfer:[IncomingTransfer](#incomingtransfer) ) ⇒</code> |
< -| [**incoming_prepare**](#event-*_prepare-) | <code>( transfer:[IncomingTransfer](#incomingtransfer) ) ⇒</code> |
< -| [**incoming_fulfill**](#event-*_fulfill-) | <code>( transfer:[IncomingTransfer](#incomingtransfer), fulfillment:String ) ⇒</code> |
< -| [**incoming_reject**](#event-*_reject-) | <code>( transfer:[IncomingTransfer](#incomingtransfer), rejectionReason:[RejectionMessage](#class-rejectionmessage) ) ⇒</code> |
< -| [**incoming_cancel**](#event-*_cancel-) | <code>( transfer:[IncomingTransfer](#incomingtransfer), cancellationReason:[RejectionMessage](#class-rejectionmessage) ) ⇒</code> |
< -| [**incoming_message**](#event-*_message-) | <code>( message:[IncomingMessage](#incomingmessage) ) ⇒</code> |
< -| [**outgoing_transfer**](#event-*_transfer-) | <code>( transfer:[outgoingTransfer](#outgoingtransfer) ) ⇒</code> |
< -| [**outgoing_prepare**](#event-*_prepare-) | <code>( transfer:[outgoingTransfer](#outgoingtransfer) ) ⇒</code> |
< -| [**outgoing_fulfill**](#event-*_fulfill-) | <code>( transfer:[outgoingTransfer](#outgoingtransfer), fulfillment:String ) ⇒</code> |
< -| [**outgoing_reject**](#event-*_reject-) | <code>( transfer:[outgoingTransfer](#outgoingtransfer), rejectionReason:[RejectionMessage](#class-rejectionmessage) ) ⇒</code> |
< -| [**outgoing_cancel**](#event-*_cancel-) | <code>( transfer:[outgoingTransfer](#outgoingtransfer), cancellationReason:[RejectionMessage](#class-rejectionmessage) ) ⇒</code> |
---
> -| [**incoming_transfer**](#event-*_transfer-) | <code>( transfer:[Transfer](#transfer) ) ⇒</code> |
> -| [**incoming_prepare**](#event-*_prepare-) | <code>( transfer:[Transfer](#transfer) ) ⇒</code> |
> -| [**incoming_fulfill**](#event-*_fulfill-) | <code>( transfer:[Transfer](#transfer), fulfillment:String ) ⇒</code> |
> -| [**incoming_reject**](#event-*_reject-) | <code>( transfer:[Transfer](#transfer), rejectionReason:[RejectionMessage](#class-rejectionmessage) ) ⇒</code> |
> -| [**incoming_cancel**](#event-*_cancel-) | <code>( transfer:[Transfer](#transfer), cancellationReason:[RejectionMessage](#class-rejectionmessage) ) ⇒</code> |
> -| [**incoming_request**](#event-*_request-) | <code>( message:[Message](#message) ) ⇒</code> |
> -| [**incoming_response**](#event-*_response-) | <code>( message:[Message](#message) ) ⇒</code> |
> -| [**outgoing_transfer**](#event-*_transfer-) | <code>( transfer:[Transfer](#transfer) ) ⇒</code> |
> -| [**outgoing_prepare**](#event-*_prepare-) | <code>( transfer:[Transfer](#transfer) ) ⇒</code> |
> -| [**outgoing_fulfill**](#event-*_fulfill-) | <code>( transfer:[Transfer](#transfer), fulfillment:String ) ⇒</code> |
> -| [**outgoing_reject**](#event-*_reject-) | <code>( transfer:[Transfer](#transfer), rejectionReason:[RejectionMessage](#class-rejectionmessage) ) ⇒</code> |
> -| [**outgoing_cancel**](#event-*_cancel-) | <code>( transfer:[Transfer](#transfer), cancellationReason:[RejectionMessage](#class-rejectionmessage) ) ⇒</code> |
> -| [**outgoing_request**](#event-*_request-) | <code>( message:[Message](#message) ) ⇒</code> |
> -| [**outgoing_response**](#event-*_response-) | <code>( message:[Message](#message) ) ⇒</code> |
42c35,36
< +| [**incoming_message**](#event-_message) | <code>( message:[IncomingMessage](#incomingmessage) ) ⇒</code> |
---
> +| [**incoming_request**](#event-_request) | <code>( message:[Message](#message) ) ⇒</code> |
> +| [**incoming_response**](#event-_response) | <code>( message:[Message](#message) ) ⇒</code> |
47a42,43
> +| [**outgoing_request**](#event-_request) | <code>( message:[Message](#message) ) ⇒</code> |
> +| [**outgoing_response**](#event-_response) | <code>( message:[Message](#message) ) ⇒</code> |
52c48
< @@ -313,7 +313,7 @@ to you.
---
> @@ -343,7 +343,7 @@ to you.
61c57
< @@ -590,9 +590,11 @@ Metadata describing the ledger. This data is returned by the [`getInfo`](#getinf
---
> @@ -638,9 +638,11 @@ Metadata describing the ledger. This data is returned by the [`getInfo`](#getinf
64,71c60,67
<  | `String` | [prefix](#prefix) | The plugin's ILP address prefix |
< -| `String` | [currencyCode](#currencyCode) | [ISO 4217](https://en.wikipedia.org/wiki/ISO_4217) three-letter currency code |
< -| `Number` | [currencyScale](#currencyScale) | Integer `(..., -2, -1, 0, 1, 2, ...)`, such that one of the ledger's base units equals `10^-<currencyScale> <currencyCode>` |
< +| `String` | [currencyCode](#currencycode) | [ISO 4217](https://en.wikipedia.org/wiki/ISO_4217) three-letter currency code |
< +| `Number` | [currencyScale](#currencyscale) | Integer `(..., -2, -1, 0, 1, 2, ...)`, such that one of the ledger's base units equals `10^-<currencyScale> <currencyCode>` |
<  | `String[]` | [connectors](#connectors) | ILP addresses of recommended connectors |
< +| `String` | [minBalance](#minbalance-optional) | Integer String, for instance `"0"`, indicating the minimum balance. Optional, defaults to zero. |
< +| `String` | [maxBalance](#maxbalance-optional) | Integer String, for instance `"1000000000000"`, indicating the maximum balance. Optional, defaults to plus infinity. |
---
>  | `String` | [prefix](#ledgerinfo-prefix) | The plugin's ILP address prefix |
> -| `String` | [currencyCode](#ledgerinfo-currencyCode) | [ISO 4217](https://en.wikipedia.org/wiki/ISO_4217) three-letter currency code |
> -| `Number` | [currencyScale](#ledgerinfo-currencyScale) | Integer `(..., -2, -1, 0, 1, 2, ...)`, such that one of the ledger's base units equals `10^-<currencyScale> <currencyCode>` |
> +| `String` | [currencyCode](#ledgerinfo-currencycode) | [ISO 4217](https://en.wikipedia.org/wiki/ISO_4217) three-letter currency code |
> +| `Number` | [currencyScale](#ledgerinfo-currencyscale) | Integer `(..., -2, -1, 0, 1, 2, ...)`, such that one of the ledger's base units equals `10^-<currencyScale> <currencyCode>` |
>  | `String[]` | [connectors](#ledgerinfo-connectors) | ILP addresses of recommended connectors |
> +| `String` | [minBalance](#ledgerinfo-minbalance-optional) | Integer String, for instance `"0"`, indicating the minimum balance. Optional, defaults to zero. |
> +| `String` | [maxBalance](#ledgerinfo-maxbalance-optional) | Integer String, for instance `"1000000000000"`, indicating the maximum balance. Optional, defaults to plus infinity. |
75c71
< @@ -624,6 +626,24 @@ currencyCode is `USD` and its currencyScale is `2`.
---
> @@ -672,6 +674,24 @@ currencyCode is `USD` and its currencyScale is `2`.
79c75
< +#### minBalance (OPTIONAL)
---
> +#### LedgerInfo#minBalance (OPTIONAL)
88c84
< +#### maxBalance (OPTIONAL)
---
> +#### LedgerInfo#maxBalance (OPTIONAL)
```

File: 0004-ledger-plugin-interface/0004-ledger-plugin-interface.md [hunk 1]
```diff
@@ -38,12 +40,15 @@ This spec depends on the [ILP spec](../0003-interledger-protocol/).
...
 | [**outgoing_transfer**](#event-_transfer) | <code>( transfer:[outgoingTransfer](#outgoingtransfer) ) ⇒</code> |
 | [**outgoing_prepare**](#event-_prepare) | <code>( transfer:[outgoingTransfer](#outgoingtransfer) ) ⇒</code> |
 | [**outgoing_fulfill**](#event-_fulfill) | <code>( transfer:[outgoingTransfer](#outgoingtransfer), fulfillment:String ) ⇒</code> |
 | [**outgoing_reject**](#event-_reject) | <code>( transfer:[outgoingTransfer](#outgoingtransfer), rejectionReason:[RejectionMessage](#class-rejectionmessage) ) ⇒</code> |
 | [**outgoing_cancel**](#event-_cancel) | <code>( transfer:[outgoingTransfer](#outgoingtransfer), cancellationReason:[RejectionMessage](#class-rejectionmessage) ) ⇒</code> |
+| [**outgoing_request**](#event-_request) | <code>( message:[Message](#message) ) ⇒</code> |
+| [**outgoing_response**](#event-_response) | <code>( message:[Message](#message) ) ⇒</code> |
 | [**info_change**](#event-info_change) | <code>( info:[LedgerInfo](#class-ledgerinfo) ) ⇒</code> |
 
 ###### Errors
```
Comment (sharafian) [inline 0004-ledger-plugin-interface/0004-ledger-plugin-interface.md]: Can you add `RequestHandlerAlreadyRegisteredError` here?

Comment (michielbdejong) [inline 0004-ledger-plugin-interface/0004-ledger-plugin-interface.md]: I don't want to edit Stefan's PR, I just want to rebase and merge it. We can do a follow-up to add this if it's missing.

Comment (sharafian) [inline 0004-ledger-plugin-interface/0004-ledger-plugin-interface.md]: ok, can you file an issue or make a follow-up PR after this is merged?

**Pull request merged**